File: python-script/FindPic.py
import cv2
import numpy as np


def to_arr(image):
    return np.array(image)[:, :, 0:3]


class CvMatchTemple:
    def __init__(self, sim=0.95, original=None, goal=None):
        self.sim = sim
        self.__original = original
        self.__goal = goal
        self.__result = None

    def similarity(self, sim):
        return CvMatchTemple(
            sim=sim,
            original=self.__original,
            goal=self.__goal)

    def original(self, original):
        return CvMatchTemple(
            sim=self.sim,
            original=original,
            goal=self.__goal)

    def goal(self, goal):
        return CvMatchTemple(
            sim=self.sim,
            original=self.__original,
            goal=goal)

    def run(self):
        assert self.__goal is not None
        assert self.__original is not None
        goal = self.__goal
        original = self.__original
        if type(goal) is np.ndarray:
            pass
        elif type(goal) is str:
            goal = cv2.imread(goal)
        else:
            goal = to_arr(goal)

        if type(original) is np.ndarray:
            pass
        elif type(original) is str:
            original = cv2.imread(original)
        else:
            original = to_arr(original)

        result = cv2.matchTemplate(goal, original, cv2.TM_CCOEFF_NORMED)
        self.__result = result
        return self

    def result(self):
        if self.__result is None:
            self.run()
        return self.__result

    def startPoint(self):
        if self.maxValue() > self.sim:
            return self.maxPoint()
        else:
            return None

    def isFind(self):
        return self.startPoint() is not None

    def noFind(self):
        return not self.isFind()

    def maxValue(self):
        result = self.result()
        return np.max(self.result())

    def maxPoint(self):
        result = self.result()
        point = np.unravel_index(result.argmax(), result.shape)
        return (int(point[1]), int(point[0]))

File: src/main/scala/utensil/FindPic.scala
package utensil

import scala.concurrent.Await
import scala.concurrent.duration.Duration

import models.{GoalImage, OriginalImage, Point}


object FindPicBuild {

  trait Image

  trait Original extends Image

  trait Goal extends Image

  trait Nothing extends Image

  type Request = Goal with Original

  def apply() = new FindPicBuild[Nothing](None, None)

  def findPic(originalName: String, goalName: String, patten: String) = {
    val result = PythonScript.eval { jep =>
      val regex = "\\(([0-9|.]+), ?([0-9]+), ?([0-9]+)\\)".r

      jep.getValue(s"jvm_find_pic('$originalName','$goalName','$patten')") match {
        case regex(sim, x, y) => (sim.toDouble, x.toInt, y.toInt)
      }
    }

    val (max, x, y) = Await.result(result, Duration.Inf)
    max -> Point(x, y)
  }

  implicit class WithRun[Arr <: Image](findPicBuild: FindPicBuild[Arr])(implicit x: Arr <:< Original with Goal) {
    def run() = {
      val original = findPicBuild.original.get
      val goal = findPicBuild.goal.get
      val patten = findPicBuild.patten
      val threshold = findPicBuild.threshold
      val originalName = original.name.replaceAll("\\\\", "/")
      val goalName = goal.name.replaceAll("\\\\", "/")

      val (similarity, topLeftPoint) = findPic(originalName, goalName, patten.toString)

      if(similarity > threshold)
        new IsFindPic(topLeftPoint, similarity)
      else
        new NoFindPic(topLeftPoint, similarity)
    }
  }

  object Patten extends Enumeration {
    val Default = Value("default")
    val Edge    = Value("edge")
  }

}

trait FindPicResult {
  def isFind: Boolean = this match {
    case IsFindPic(_) => true
    case NoFindPic()  => false
  }

  def similarity = this match {
    case x@IsFindPic(_) => x.sim
    case x@NoFindPic()  => x.sim
  }

  def noFind = !isFind

  def point = this match {
    case IsFindPic(x) => x
    case _            => ???
  }
}

class IsFindPic(val topLeftPoint: Point, val sim: Double) extends FindPicResult

object IsFindPic {
  def unapply(arg: IsFindPic): Option[Point] = Some(arg.topLeftPoint)
}

class NoFindPic(val topLeftPoint: Point, val sim: Double) extends FindPicResult

object NoFindPic {
  def unapply(arg: NoFindPic) = true
}

class FindPicBuild[+Arr] private(val original: Option[OriginalImage],
                                 val goal: Option[GoalImage],
                                 val threshold: Double = 0.95,
                                 val patten: FindPicBuild.Patten.Value = FindPicBuild.Patten.Default) {

  def withGoal(goal: GoalImage) =
    new FindPicBuild[Arr with FindPicBuild.Goal](goal = Some(goal), original = original, threshold = threshold, patten = patten)

  def withOriginal(original: OriginalImage) =
    new FindPicBuild[Arr with FindPicBuild.Original](goal = goal, original = Some(original), threshold = threshold, patten = patten)

  def withThreshold(threshold: Double) =
    new FindPicBuild[Arr](goal = goal, original = original, threshold = threshold, patten = patten)

  def withPatten(threshold: Double) =
    new FindPicBuild[Arr](goal = goal, original = original, threshold = threshold, patten = patten)

}


//case class FindPic(original: OriginalImage,
//                   goal: GoalImage,
//                   threshold: Double = 0.95,
//                   patten: FindPic.Patten.Value = FindPic.Patten.Default) {
//  lazy val (similarity, topLeftPoint) = {
//    val originalName = original.name.replaceAll("\\\\", "/")
//    val goalName = goal.name.replaceAll("\\\\", "/")
//
//    val result = PythonScript.eval { jep =>
//      val regex = "\\(([0-9|.]+), ?([0-9]+), ?([0-9]+)\\)".r
//
//      jep.getValue(s"jvm_find_pic('$originalName','$goalName','$patten')") match {
//        case regex(sim, x, y) => (sim.toDouble, x.toInt, y.toInt)
//      }
//    }
//
//    val (max, x, y) = Await.result(result, Duration.Inf)
//    max -> Point(x, y)
//  }
//
//  def point: Option[Point] = if(isFind) Some(topLeftPoint) else None
//
//  def isFind: Boolean = similarity >= threshold
//
//  def noFind: Boolean = !isFind
//}

